src/compiler_lectures/how_salsa_works.md

```diff
@@ -0,0 +1,169 @@
+
+This means that when you try to compute with the library, Salsa is going to compute various derived values, and eventually read the input and produce the result for the asked computation.
+
+Salsa is going to track, in the course of computing, which inputs were accessed, which derived values, and this is going to be used later to determine what's going to happen when the inputs change: are the derived values still valid? 
+
+This doesn't mean necessarily that each computation downstream from the input is going to be checked, since that is going to be costly. Salsa only needs to check each downstream computation until it finds one that isn't changed, therefore it won't need to check the other derived computations, since they wouldn't need to change.
+
+It's is helpful to think about this as a graph with nodes. Each derived value has a dependency on some other values, that could be base or derived. Base values don't have a dependency.
+
+```ignore
+I <- A <- C ...
+          |
+J <- B <--+
+```
+
+When an input `I` changes, the derived value `A` could have changed, but another derived value `B` , which doesn't depend neither on `I`, nor does it depends on `A`, nor does it depends on any other derived value from `A` or `I`, is not subject to change.
+Therefore, Salsa can reuse the computation done for `B` in the past, without having to compute it again.
+
+The computation could also terminate early. Keeping the same graph as before, say that input `I` has changed in some way (and input `J` hasn't) but, when computing `A` again, it's found that `A` hasn't changed from the previous computation. This leads to an "early termination", because there's no need to check if `C` needs to change, since both `C` direct inputs, `A` and `B`, haven't changed.
+
+## Key Salsa concepts
+### Query
+A query is some value that Salsa can access in the course of computation.
+Each query can have a number of keys (from 0 to many), and all queries have a result, akin to functions.
+0-key queries are called "input" queries.
+### Database
+The database is basically the context for the entire computation, it's gonna store all Salsa's internal state, all intermidiate values for each query, and anything else that the computation might need.
+The database to know all the queries that the library is going to do before it can be build, but they don't need to be specified in the same place.
+
+After the database is formed, it can be accessed with queries that are basically going to work like functions.
+Since each query is going to be stored in the database, when a query is invoked N times, it's going to return N **cloned** results, without having to recompute the query (unless the input has changed in such a way that it warrants recomputation).
+
+For each input query (0-key), there's going to be a "set" method, that allows to change the output of such query, and trigger previous memoized values to be potentially invalidated.
```

Comment:
All of the "going to be" phrasing makes it sound like these are features that Salsa could have in the future, but doesn't currently. I'd review the text to phrase it in the present tense.

Reply:
If you are talking about this section specifically, it's because this refers to something that's going to happen when the library is used... If talking about the whole article: the text is based of a video done a while ago, some stuff might not had yet been implemented by that point. I don't know what's in salsa now so I can't really tell if something should be on the present tense. Either way, I decided to rewrite this section a bit with present tense as you suggested